feat: support additional services for blue-green. fixes #451

d3adb5 opened the pull request:
Add support for additional preview and active services for the BlueGreen
strategy.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

This is my first contribution to Argo Rollouts, and I feel like #451 isn't
really closed if this doesn't encompass the Canary strategy. I want to work on
it, but first I wanted to write E2E tests for this feature. I'm rather new to
Go, so I'm not sure if I'm doing this right. Still combing through the available
tests to see if I can find a good example to follow.

Support for multiple services is added as a couple new optional fields called
`additionalPreviewServices` and `additionalActiveServices`. When populated with
service names, these lists will cause the services in them to be updated as
though they were _the_ preview service or _the_ active service.

Help with this would be much appreciated.

meeech commented:
Would it make sense to add activeServices/previewServices, and just make them mutually exclusive?

d3adb5 commented:
> Would it make sense to add activeServices/previewServices, and just make them mutually exclusive?

If I understand your suggestion correctly, I don't think so. My original use case was I had two `Service`s for a specific service: one attached to a Network Load Balancer, redirecting UDP traffic, and a simple ClusterIP referenced by `Ingress`. There were two active and two preview `Service`s.

There are more use cases reported in #451. The problem with just using a list is it's not explicit which `Service` might be the "origin of the truth" in the case of recovering from a crash. Not that that behavior is really made explicit here.

meeech commented:
@d3adb5  Sorry, i realize re-reading it that I wasn't clear there. I didn't mean to make them mutually exclusive to each other (i know you need active and preview). 
i meant to make the plural and singular (eg: activeService and activeServices) mutually exclusive.  
Which I now realize is similar to the suggestion @huikang  https://github.com/argoproj/argo-rollouts/issues/451#issuecomment-933017263 made on #451 - though i don't know if they meant to suggest making them mutually exclusive (basically, can only use 1 - the singular, or the plural) with the plural being preferred going forward - since it works with 1 or more services.

As far as source of truth, i would make the first in the list the SoT, and document that.

d3adb5 commented:
I realize it's been a while since I touched this PR. Life's been busy, now there are even merge conflicts. I'll get to it sooner or later.

> As far as source of truth, i would make the first in the list the SoT, and document that.

My goal with the `additional*` parameters was to avoid having to document which one will be the SoT. My hopes were reading the spec would make it immediately clear which one would be used in case of recovery and which ones would become secondary updates, rather than let the user guess which one would be the SoT and have to read the documentation to confirm their assumptions.

fdfzcq commented:
Hello @d3adb5 any updates on this? :) We are in need of the same feature and wonder if we can help contribute and push for the feature.

d3adb5 commented:
Hey @fdfzcq, I haven't had the time to look back at this, and since I've moved on from the project where the use case came up it took a hit in my list of priorities. Last I remember I was a little confused about how to add E2E tests for this and still wondering if I'm going about the type change the right way.

I have a rough week or two ahead of me at work right now, but afterwards I'll be able to commit myself to this. If you want, you can always take over this implementation. Want to explain your use case here or elsewhere so I get a clearer picture of what you guys need?

eduardOrthopy commented:
So, it has been a while since there were changes made here. @d3adb5 if you are still on this, please let me/us know if/how I/we can support.

This is a feature that I would really love to see in Argo, and that would solve a big pain point for me.

If you cannot find time to contribute to this any more, that is fair. Life moves on and so do we. Just let us know, so that the rest of the community can take over. And thanks for creating this PR in the first place.

d3adb5 committed:
feat: support additional services for blue-green

Add support for additional preview and active services for the BlueGreen
strategy.

Signed-off-by: d3adb5 <me@d3adb5.net>

d3adb5 commented:
Sorry for the delay getting back to everyone. @eduardOrthopy, it would help me quite a bit to understand how to properly run and write E2E tests for this. For some reason E2E tests with a kind or k3d cluster on my machine are always failing intermittently. I'm new to Go, too, so any feedback on what I have here would be well appreciated.

I rebased on `master` and solved conflicts, ran `make codegen` again, and will be looking into this during my spare time. Please add as much feedback as you like here! If anybody wants to push to my branch, I'm fine with that too. :thinking:

eduardOrthopy commented:
@d3adb5 sorry for the delay on my side this time. As I said, it is open source, and you have a life, as do all others that work on this. Priorities change over time.

Thanks for letting the community know how you feel about joint work on this topic.

I will check out the branch and look at the IT E2E testing part. Maybe I can get some clarity into that.

yvmr-33 commented:
@eduardOrthopy @d3adb5 
Hello, Is this feature available now or yet to be done?
My use case is, we have 2 services for a single deployment, how to do weighted traffic routing for both the services?

eduardOrthopy commented:
@yvmr-33 I have had a lot of shift in priorities and could not invest more than a half day into this in the last year. So from my end, sorry, no forward motion.

Would appreciate it if you could look at this/take over/contribute, as I do not see this on my roadmap for the foreseeable future.

insylogo commented:
This is obviously a very valuable feature that lots of people have interest in. If the core maintainers could find it inthemselves to push this over the line, I'd love it too.

insylogo reviewed (approved):
Haven't seen the result of CI build processes but looks good to me in this state

rideliner commented:
You also need to add the additional services in `GetRolloutServiceKeys` or else reconciling one of the additional services won't get the rollout's targets.

```diff
diff --git a/utils/service/service.go b/utils/service/service.go
index fbcfb63a..4760630f 100644
--- a/utils/service/service.go
+++ b/utils/service/service.go
@@ -25,9 +25,15 @@ func GetRolloutServiceKeys(rollout *v1alpha1.Rollout) []string {
 		if rollout.Spec.Strategy.BlueGreen.ActiveService != "" {
 			services = append(services, fmt.Sprintf("%s/%s", rollout.Namespace, rollout.Spec.Strategy.BlueGreen.ActiveService))
 		}
+		for _, svc := range rollout.Spec.Strategy.BlueGreen.AdditionalActiveServices {
+			services = append(services, fmt.Sprintf("%s/%s", rollout.Namespace, svc))
+		}
 		if rollout.Spec.Strategy.BlueGreen.PreviewService != "" {
 			services = append(services, fmt.Sprintf("%s/%s", rollout.Namespace, rollout.Spec.Strategy.BlueGreen.PreviewService))
 		}
+		for _, svc := range rollout.Spec.Strategy.BlueGreen.AdditionalPreviewServices {
+			services = append(services, fmt.Sprintf("%s/%s", rollout.Namespace, svc))
+		}
 	} else if rollout.Spec.Strategy.Canary != nil {
 		if rollout.Spec.Strategy.Canary.CanaryService != "" {
 			services = append(services, fmt.Sprintf("%s/%s", rollout.Namespace, rollout.Spec.Strategy.Canary.CanaryService))
```

d3adb5 commented:
Applied your patch, @rideliner, thank you! Also rebase it on `master`. I'll flag this as a ready for review since I'd like us to get the ball rolling. I still would love some help implementing E2E tests for this.

Thank you all for your involvement and attention.

d3adb5 committed:
Merge branch 'master' into feat/multiple-svc
